composer/utils/checkpoint.py

```diff
@@ -443,23 +453,75 @@ def __init__(self, source_path: str, destination_path: str, object_store):
             super().__init__(destination_path)
 
         def read_data(self, plan: LoadPlan, planner: LoadPlanner):
-            # 1. Download to the destination all files that this rank is responsible for.
-            for plan_item in plan.items:
-                # Each plan item has a storage index which points to the relative path of the shard file at save time.
-                relative_file_path = self.storage_data[plan_item.storage_index].relative_path
-                # Download the shard file to the relative path it's associated to and save that relative path
-                # to the root directory specified to the FileSystem reader constructor.
-                file_destination = str(Path(self.destination_path) / Path(relative_file_path))
-                # The file could have already been downloaded as diffeent plan items can point to same file.
-                if not os.path.exists(file_destination):
-                    self.object_store.download_object(object_name=str(
-                        Path(self.source_path) / Path(relative_file_path)),
-                                                      filename=file_destination)
+            device_mesh = state.fsdp_device_mesh
+            first_replica = device_mesh is None or device_mesh.get_local_rank(mesh_dim=0) == 0
+
+            # 1. Download to the destination all files this rank needs if on first replica
+            if first_replica:
+                log.debug(f'Rank {dist.get_global_rank()} starting to download files.')
+                for plan_item in plan.items:
+                    # Each plan item has a storage index which points to the relative path of the shard file at save time.
+                    relative_file_path = self.storage_data[plan_item.storage_index].relative_path
+                    # Download the shard file to the relative path it's associated to and save that relative path
+                    # to the root directory specified to the FileSystem reader constructor.
+                    file_destination = str(Path(self.destination_path) / Path(relative_file_path))
+                    # The file could have already been downloaded as different plan items can point to same file.
+                    if not os.path.exists(file_destination):
+                        log.debug(f'Downloading {relative_file_path} to {file_destination}.')
+                        self.object_store.download_object(object_name=str(
+                            Path(self.source_path) / Path(relative_file_path)),
+                                                          filename=file_destination)
+                        log.debug(f'Finished downloading {relative_file_path} to {file_destination}.')
 
             # 2. Wait for all ranks to finish.
+            log.debug(f'Finished downloading all files.')
             dist.barrier()
```

Comment:
If the single replica download takes longer than `dist_timeout`, this will nccl out right?

Reply:
Yes. It's pretty fast though (<2 min at 100B+ scale)

Reply:
This depends on your model size vs replica size though, so definitely could happen...maybe just jira because we don't have a mechanism for cross node busy wait.